### PR TITLE
Adds restart capabilities to control-duck channel

### PR DESCRIPTION
**Pull request by prestonraab**

### CODE REVIEW

File: discord_bot2.py [hunk 1]
```diff
@@ -81,7 +82,7 @@ def __init__(self, prompt_dir: Path, conversation_dir: Path):
 
     def _load_prompts(self, prompt_dir: Path):
         self.prompts = {}
-        for file in prompt_dir.iterdir():
+        for file in prompt_dir.glob("**/*"):
```
Comment (byubean) [inline discord_bot2.py:85]: What if two files in different folders have the same name? 

I would keep the original behavior for now.

File: discord_bot2.py [hunk 2]
```diff
@@ -163,8 +164,105 @@ async def on_ready(self):
...
+        return
+
+    async def execute_message(self, message):
+        """
+        Execute a command in the shell and return the output to the channel
+        Removes the starting !
+        :param message: discord.Message object, used to get the command
+        """
+        await self.execute_command(message.content[1:], message.channel)
+        return
```
Comment (byubean) [inline discord_bot2.py:227]: This method seems like a security nightmare waiting to happen. It doesn't look like it is used though (which is good!). Is it left over from before?

File: discord_bot2.py [hunk 3]
```diff
@@ -163,8 +164,105 @@ async def on_ready(self):
...
+        :param message:
+        :return:
+        """
+        content = message.content
+        if content.startswith('!restart'):
+            await self.restart(message)
+
+        elif content.startswith('!log'):
+            await self.execute_command("cat /tmp/duck.log", message.channel)
+            await message.channel.send('Done.')
```
Comment (byubean) [inline discord_bot2.py:250]: The log file can get quite large. Perhaps we have the duck share the file instead of write the content?

File: discord_bot2.py [hunk 4]
```diff
@@ -180,9 +278,16 @@ async def on_message(self, message):
         if message.content.startswith('//'):
             return
 
+        if message.channel.name == 'control-duck':
+            await self.control_on_message(message)
+            return
+
```
Comment (byubean) [inline discord_bot2.py:284]: I want something like 
```python
if message.channel.id in self.control_channels:
```
And then we configure the control channels in a configuration file. 

I want to use `id` instead of `name` so we can specify the exact channels on the exact servers. This way admins of other course servers can't simply add a channel named `control-duck` and take over.

File: discord_bot2.py [hunk 5]
```diff
@@ -180,9 +278,16 @@ async def on_message(self, message):
...
         # if the message is in a listen channel, create a thread
         if message.channel.name in self.prompts:
-            prefix = self.prompts[message.channel.name]
+            prefix = ""
+            if message.channel.category.name.lower() in self.prompts:
+                prefix += self.prompts[message.channel.category.name.lower()]
+            prefix += self.prompts[message.channel.name]
```
Comment (byubean) [inline discord_bot2.py:290]: I think this is a clever idea, but not quite how I would recommend doing it. I would take this part out of this PR and do a separate PR to address hierarchical prompts.

File: discord_bot2.py [hunk 6]
```diff
@@ -163,8 +164,105 @@ async def on_ready(self):
...
+        :param channel: The channel to say something in
+        :param output: The message to say, long messages will be split into multiple messages
+        """
+        # Async methods like for loops better than while loops
+        # Split output into 1800 character chunks
+        for i in range(len(text) // 1800):
+            await channel.send(f'{prefix}{text[:1800]}{suffix}')
+            text = text[1800:]
+
+        await channel.send(f'{prefix}{text}{suffix}')
```
Comment (byubean) [inline discord_bot2.py:200]: Isn't this logic already present somewhere? Why rewrite it?

File: discord_bot2.py [hunk 7]
```diff
@@ -163,8 +164,105 @@ async def on_ready(self):
...
+                await channel.send('Duck online')
 
-    async def on_message(self, message):
+    async def restart(self, message):
+        """
+        Restart the bot
+        :param message: The message that triggered the restart
+        """
+        await message.channel.send(f'Restart requested.')
+        os.chdir(Path(__file__).parent)
```
Comment (byubean) [inline discord_bot2.py:178]: I would move this logic to `execute_command`. In `execute_command`:
```python
work_dir = Path(__file__).parent
subprocess.run(..., cwd=work_dir)
```

File: discord_bot2.py (hunk 6 above)
Comment (prestonraab) [reply discord_bot2.py]: Sorry about that, I didn't realize that's what `parse_blocks` did. Since `send` currently uses `parse_blocks` but takes a thread, I think I'll make a similar `send_in_channel`.

File: discord_bot2.py [hunk 8]
```diff
@@ -163,8 +164,105 @@ async def on_ready(self):
...
+        Execute a command in the shell and return the output to the channel
+        """
+        # Run command using shell and pipe output to channel
+        await self.say_in_channel(channel, text, "```ps\n$ ", "```")
+        process = subprocess.run(text, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        # Get output of command and send to channel
+        errors = process.stderr.decode('utf-8')
+        if errors:
+            await self.say_in_channel(channel, errors, prefix=f'Errors: ```', suffix='```')
+        output = str(process.stdout.decode('utf-8'))
```
[the remark this answers is not in the document]
Comment (prestonraab) [reply discord_bot2.py]: After testing, apparently not :)

File: discord_bot2.py (hunk 4 above)
Comment (prestonraab) [reply discord_bot2.py]: Something like this?
config.json:
{
  "control_channels": [1111389107587928154]
}

File: discord_bot2.py (hunk 3 above)
Comment (prestonraab) [reply discord_bot2.py]: Good idea!

File: discord_bot2.py [hunk 9]
```diff
@@ -67,6 +68,152 @@ def from_json(jobj: dict, thread: discord.Thread) -> 'Conversation':
...
+
+async def send(thread: discord.Thread, text: str):
+    for block in parse_blocks(text):
+        await thread.send(block)
+
+
+async def send_in_channel(channel: discord.TextChannel, text: str):
+    for block in parse_blocks(text):
+        await channel.send(block)
+
```
Comment (byubean) [inline discord_bot2.py:196]: This is identical to `send`, so I don't think we need it.

If you want to get the typing right, use `Union[discord.Thread, discord.TextChannel]`

File: discord_bot2.py [hunk 10]
```diff
@@ -163,8 +289,12 @@ async def on_ready(self):
         logging.info(self.user.name)
         logging.info(self.user.id)
         logging.info('------')
+        channels = self.get_all_channels()
+        for channel in channels:
+            if channel.name == 'control-duck':
+                await channel.send('Duck online')
```
Comment (byubean) [inline discord_bot2.py:295]: Instead, look up the control channels.

File: discord_bot2.py [hunk 11]
```diff
@@ -180,14 +310,20 @@ async def on_message(self, message):
         if message.content.startswith('//'):
             return
 
+        if message.channel.id in self.control_channels:
+            await control_on_message(message)
+            return
```
Comment (byubean) [inline discord_bot2.py:315]: missing extra blank line after `return`

File: discord_bot2.py [hunk 12]
```diff
@@ -180,14 +310,20 @@ async def on_message(self, message):
...
+        elif message.channel.name == 'control-duck':
+            await send_in_channel(message.channel, f'Channel id: {message.channel.id}')
+            return
```
Comment (byubean) [inline discord_bot2.py:318]: Still need to check if channel ID is in the control channels.

File: config.json [hunk 13]
```diff
@@ -0,0 +1,3 @@
+{
+  "control_channels": [1069014613779300524]
```
Comment (byubean) [inline config.json:2]: Is this the control channel in the 110 server?

File: discord_bot2.py [hunk 14]
```diff
@@ -67,6 +68,149 @@ def from_json(jobj: dict, thread: discord.Thread) -> 'Conversation':
...
+    This function is called whenever the bot sees a message in a control channel
+    :param message:
+    :return:
+    """
+    content = message.content
+    if content.startswith('!restart'):
+        await restart(message)
+
+    elif content.startswith('!log'):
+        await message.channel.send(file=discord.File('/tmp/duck.log'))
```
Comment (byubean) [inline discord_bot2.py:126]: Future PR: configure the Logger to write to the log file. Configure the log file path in the command-line args, then use that path here too.

File: discord_bot2.py [hunk 15]
```diff
@@ -67,6 +68,149 @@ def from_json(jobj: dict, thread: discord.Thread) -> 'Conversation':
...
+
+    elif content.startswith('!rmlog'):
+        await execute_command("rm /tmp/duck.log", message.channel)
+        await execute_command("touch /tmp/duck.log", message.channel)
+
+    elif content.startswith('!status'):
+        await message.channel.send('I am alive.')
```
Comment (byubean) [inline discord_bot2.py:133]:
```python
message.channel.send('🦆')
```
Future PR. :)

File: restart.sh [hunk 16]
```diff
@@ -1,2 +1,2 @@
-kill `ps -ae | grep python | cut -d ' ' -f 1`
+kill "$(pgrep -a python | awk '{$1=$1};1' | cut -d ' ' -f 1)"
```
Comment (byubean) [inline restart.sh:1]: why pgrep + awk instead of ps + grep?

Comment (prestonraab) [reply restart.sh:1]: awk is used to make the command platform independent: while Macs put one space at the start of ps, Linux puts 3 and Windows puts 4

pgrep was suggested by Max's Pycharm as an alternative to ps + grep

File: config.json (hunk 13 above)
Comment (prestonraab) [reply config.json:2]: Yes

File: pyproject.toml [hunk 17]
```diff
@@ -8,6 +8,7 @@ authors = ["Gordon Bean <gbean@cs.byu.edu>"]
 python = "^3.9"
 openai = "^0.27.2"
 discord = "^2.1.0"
+quest = "^0.1.3"
```
Comment (byubean) [inline pyproject.toml:11]: We should remove this in this PR.

**Review by byubean — APPROVED** [review]
